cmd/shell.go

```diff
@@ -305,9 +314,110 @@ func ReadConfigFile(filename string, out interface{}) error {
 	}
 	defer file.Close()
 
-	decoder := json.NewDecoder(file)
-	decoder.DisallowUnknownFields()
-	return decoder.Decode(out)
+	return decodeJSONStrict(file, out)
+}
+
+// ReadAndValidateConfigFile takes a file path as an argument and attempts to
+// unmarshal the content of the file into a struct containing a configuration of
+// a boulder component specified by name (e.g. boulder-ca, bad-key-revoker,
+// etc.). Any config keys in the JSON file which do not correspond to expected
+// keys in the config struct will result in errors. It also validates the config
+// using the struct tags defined in the config struct.
+func ReadAndValidateConfigFile(name, filename string) error {
+	file, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	if name == "boulder-observer" {
+		// Only the boulder-observer uses YAML config files.
+		return ValidateYAMLConfigByComponent(name, file)
+	}
+	return ValidateJSONConfigForComponent(name, file)
+}
+
+// ValidateJSONConfigForComponent takes a component name (e.g. boulder-ca
+// bad-key-revoker, etc.) and an io.Reader containing a JSON representation of a
+// config and validates the config using the struct tags defined in the config
+// struct.
+func ValidateJSONConfigForComponent(name string, in io.Reader) error {
```

Comment:
I don't believe this function is called by any other packages (since boulder/main.go calls ReadAndValidateConfigFile), so it doesn't need to be exported. The same is true for the YAML function below.

That also means that they can both take a Validator as their first argument, consolidating the lookup and construction of the Validator into ReadAndValidateConfigFile, and simplifying unit test setup (because you only have to construct a Validator to pass into this function, rather than having to construct a registry).

Speaking of which, although there are the "unitegration" tests which validate our integration test json configs in cmd/boulder/main_test.go, I think there should be true unit tests for these functions in cmd/shell_test.go as well. These would be particularly useful for testing the "unhappy path", by creating tiny minimal-reproduction configs and json strings that _fail_ validation.

Reply:
> Speaking of which, although there are the "unitegration" tests which validate our integration test json configs in cmd/boulder/main_test.go, I think there should be true unit tests for these functions in cmd/shell_test.go as well. These would be particularly useful for testing the "unhappy path", by creating tiny minimal-reproduction configs and json strings that fail validation.

I could be misunderstanding you a bit, but wouldn't this just result in us unit testing that the validator library works? I could see a test which loads a config that breaks all the rules for each of our components. However, such a task would take quite a bit of time, and in the end feels a lot like we're still testing the library.

Reply:
I guess it mostly just seems weird to expose a three-function public API, where one of those three just delegates to the other two. Like, taking it in the other direction, `ReadAndValidateConfigFIle` could just be implemented directly inside cmd/boulder/main.go, since that is the only caller and its functionality doesn't have anything to do with validation specifically.

These helper functions do a lot more than just invoke the validator library: they have error return paths if the config name isn't recognized, they _could_ have an error return if `validate.RegisterValidation()` returns an error but don't currently check that error, they return an error if json decoding fails, and they process any validation errors into a human-readable form. It's worth testing those extra paths.

Reply:
> I guess it mostly just seems weird to expose a three-function public API, where one of those three just delegates to the other two. Like, taking it in the other direction, ReadAndValidateConfigFIle could just be implemented directly inside cmd/boulder/main.go, since that is the only caller and its functionality doesn't have anything to do with validation specifically.

I actually really like moving `ReadAndValidateConfigFile()` to `boulder/main.go`. Thanks!

Reply:
> These helper functions do a lot more than just invoke the validator library: they have error return paths if the config name isn't recognized, they could have an error return if validate.RegisterValidation() returns an error but don't currently check that error, they return an error if json decoding fails, and they process any validation errors into a human-readable form. It's worth testing those extra paths.

Fair, I've gone ahead and added some unhappy path testing. Let me know if I've got it right.